elfpy/markets/hyperdrive/hyperdrive_pricing_model.py

```diff
@@ -442,3 +448,282 @@ def calc_tokens_out_given_lp_in(
+            share_reserves, bond_reserves, ONE_18 - time_stretch, share_price, initial_share_price
+        )
+        if share_reserves + dz >= (longs_outstanding + dy) / share_price:
+            return MaxLongResult(base_amount=dz * share_price, bond_amount=dy)
+
+        # To make an initial guess for the iterative approximation, we consider
+        # the solvency check to be the error that we want to reduce. The amount
+        # the long buffer exceeds the share reserves is given by
+        # (y_l + dy) / c - (z + dz). Since the error could be large, we'll use
+        # the realized price of the trade instead of the spot price to
+        # approximate the change in trade output. This gives us dy = c * 1/p * dz.
+        # Substituting this into error equation and setting the error equal to
+        # zero allows us to solve for the initial guess as:
+        #
+        # (y_l + c * 1/p * dz) / c - (z + dz) = 0
+        #              =>
+        # (1/p - 1) * dz = z - y_l/c
+        #              =>
+        # dz = (z - y_l/c) * (p / (p - 1))
+        p = share_price * dz / dy
+        dz = (share_reserves - longs_outstanding / share_price) * p / (ONE_18 - p)
+        dy = calculate_bonds_out_given_shares_in(
+            share_reserves, bond_reserves, dz, ONE_18 - time_stretch, share_price, initial_share_price
+        )
+
+        result = MaxLongResult(base_amount=FixedPoint(), bond_amount=FixedPoint())
+
+        # Our maximum long will be the largest trade size that doesn't fail
+        # the solvency check.
+        for _ in range(max_iterations):
+            # Even though YieldSpace isn't linear, we can use a linear approximation
+            # to get closer to the optimal solution. Our guess should bring us close
+            # enough to the optimal point that we can linearly approximate the
+            # change in error using the current spot price.
+            #
+            # We can approximate the change in the trade output with respect to
+            # trade size as dy' = c * (1/p) * dz'. Substituting this into our error
+            # equation and setting the error equation equal to zero allows us to
+            # solve for the trade size update:
+            #
+            # (y_l + dy + c * (1/p) * dz') / c - (z + dz + dz') = 0
+            #                  =>
+            # (1/p - 1) * dz' = (z + dz) - (y_l + dy) / c
+            #                  =>
+            # dz' = ((z + dz) - (y_l + dy) / c) * (p / (p - 1)).
+            p = calculate_spot_price(share_reserves + dz, bond_reserves - dy, initial_share_price, time_stretch)
+            error = int((share_reserves + dz) - (longs_outstanding + dy) / share_price)
+            if error > 0 and dz * share_price > result.base_amount:
+                result = MaxLongResult(base_amount=dz * share_price, bond_amount=dy)
+            if p >= ONE_18:
+                break
+            if error < 0:
+                dz -= -error * p / (ONE_18 - p)
```

Comment:
Sol code clips this against zero. You might want to do the same:
`dz = FixedPointMath.maximum(0, dz - (-error * p / (ONE_18 - p)))`

Reply:
```
                uint256 delta = uint256(-approximationError).mulDivDown(
                    p,
                    FixedPointMath.ONE_18 - p
                );
                if (dz > delta) {
                    dz -= delta;
                } else {
                    dz = 0;
                }
```